# 3.构建岗位匹配度和求职者满意度的模型

### 根据问题2的招聘信息与求职者信息，构建岗位匹配度和求职者满意度的模型，基于该模型，为每条招聘信息提供岗位匹配度非0的求职者，将结果进行降序排序存放在“resul3-1.csv”文件中，以及为每位求职者提供求职者满意度非0的招聘信息，将结果进行降序排序存放在“result3-2.csv”文件中

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib

"""
font:设置中文
unicode_minus:显示负号
"""
matplotlib.rcParams['font.family'] = ['SimHei']
matplotlib.rcParams['axes.unicode_minus']=False     # 正常显示负号


# 使用dataframe行列对齐
pd.set_option('display.unicode.ambiguous_as_wide', True)
pd.set_option('display.unicode.east_asian_width', True)

In [2]:
look_work = pd.read_excel('F:/amyself/c题爬虫/第一题/找工作.xls')
look_ability = pd.read_excel('F:/amyself/c题爬虫/第一题/找人才.xls')
look_work = pd.DataFrame(look_work)
look_ability = pd.DataFrame(look_ability)

In [3]:
look_work.head(2)

工作id      工作更新时间    工作岗位名称  工作岗位最低工资  \
0  1631112859985509888  2023-03-02 10:03      会计实习生              3500   
1  1631112859897430016  2023-03-02 10:03  技术服务工程师              5000   

   工作岗位最高工资 工作岗位要求经验  工作岗位要求学历  工作岗位招收人数  \
0              7000             不限                 2                 2   
1             10000             不限                 2                 5   

                                                           工作岗位具体地址  \
0  广东省深圳市宝安区西乡街道南昌社区固戍二路99号(南昌第二工业区东区)......   
1  广东省深圳市宝安区西乡街道南昌社区固戍二路99号(南昌第二工业区东区)......   

       公司名称           公司所属类型 公司成立模式   公司规模 岗位技能  
0  海柔创新科技  ["电子商务","互联网"]         合资  150-500人     None  
1  海柔创新科技  ["电子商务","互联网"]         合资  150-500人     None

In [4]:
look_ability.head(2)

人才id    姓名  性别 工作经验  \
0  1635610357617786880  欧先生   0.0   无经验   
1  1634019092572798976  赖女士   1.0   无经验   

                                   应聘工作岗位  最低薪资期望  最高薪资期望  \
0  ["数据分析师","数据挖掘工程师","算法工程师"]          3000          4000   
1                                ["数据分析师"]          4000          5000   

                       所在城市    上传时间 个人技能  
0  ["广东省","广州市","天河区"]  2023-03-14     None  
1  ["江苏省","南京市","江宁区"]  2023-03-10     None

In [5]:
a = ['工作id','工作岗位名称','工作岗位最低工资','工作岗位要求经验','岗位技能']
aa = look_work[a]
aa = pd.DataFrame(aa)
b = ['人才id','应聘工作岗位','最低薪资期望','工作经验','个人技能']
bb = look_ability[b]
bb = pd.DataFrame(bb)
bb['工作经验'] = bb['工作经验'].replace(np.nan, '无经验')
cc = pd.concat([aa,bb],axis=1)
cc.isnull().sum()

工作id              9334
工作岗位名称        9334
工作岗位最低工资    9334
工作岗位要求经验    9334
岗位技能            9334
人才id                 0
应聘工作岗位           0
最低薪资期望           0
工作经验               0
个人技能               0
dtype: int64

In [6]:
cc = cc.dropna()
cc

工作id        工作岗位名称  工作岗位最低工资 工作岗位要求经验  \
0     1.631113e+18          会计实习生            3500.0             不限   
1     1.631113e+18      技术服务工程师            5000.0             不限   
2     1.629825e+18  大数据分析师（BI）            4500.0             不限   
3     1.613440e+18  自然语言处理工程师            2000.0         经验不限   
4     1.613440e+18          爬虫工程师            2000.0         经验不限   
...            ...                 ...               ...              ...   
1569  1.463032e+18      前端开发工程师            8000.0         经验不限   
1570  1.374181e+18      软件测试工程师            6000.0            1-3年   
1571  1.374177e+18      新媒体运营助理            5000.0            1-3年   
1572  1.374178e+18        数据产品经理          150000.0            3-5年   
1573  1.374197e+18      数据分析实习生             100.0         经验不限   

                                        岗位技能               人才id  \
0                                           None  1635610357617786880   
1                                           None  1634019092572798976   
2                                           None  1632947816114552832   
3                               算法研究深度学习  1632286586903330816   
4                         Python爬虫框架数据采集  1579833809761861632   
...                                          ...                  ...   
1569  数据查询数据计算数据挖掘数据分析数据可视化  7535876391727459860   
1570                        系统测试软件系统搭建  7535876400317394452   
1571                                    产品运营  7535876404612361748   
1572  数据探索数据可视化项目管理需求分析产品设计  7535876413202296340   
1573                    数据探索数据查询数据建模  7535876417497263636   

                                          应聘工作岗位  最低薪资期望  \
0         ["数据分析师","数据挖掘工程师","算法工程师"]          3000   
1                                       ["数据分析师"]          4000   
2     ["数据分析师","数据挖掘工程师","机器学习工程师"]          6000   
3     ["数据分析师","数据挖掘工程师","机器学习工程师"]          4000   
4                      ["数据分析师","数据挖掘工程师"]          4000   
...                                                ...           ...   
1569                   ["数据分析师","数据挖掘工程师"]          4000   
1570                   ["数据分析师","数据挖掘工程师"]          4000   
1571                   ["数据分析师","数据挖掘工程师"]          4000   
1572                   ["数据分析师","数据挖掘工程师"]          4000   
1573                   ["数据分析师","数据挖掘工程师"]          4000   

         工作经验 个人技能  
0          无经验     None  
1          无经验     None  
2     1年工作经验     None  
3          无经验     None  
4          无经验     None  
...           ...      ...  
1569       无经验     None  
1570       无经验     None  
1571       无经验     None  
1572       无经验     None  
1573       无经验     None  

[1574 rows x 10 columns]

### 文本匹配训练模型

In [7]:
cc['工作岗位要求经验'].unique()

array(['不限', '经验不限', '5-10年', '1-3年', '3-5年', '5-7年', '7年以上', '3', '5',
       '1', '0', '10'], dtype=object)

In [8]:
cc['工作经验'].unique()

array(['无经验', '1年工作经验', '4年工作经验', '3年工作经验', '10年以上工作经验'], dtype=object)

In [9]:
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace(np.nan, '经验不限')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('不限', '经验不限')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('0', '经验不限')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('1', '1年工作经验')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('5-10年', '5年工作经验')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('1-3年', '1年工作经验')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('3-5年', '3年工作经验')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('5-7年', '5年工作经验')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('7年以上', '7年工作经验')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('3', '3年工作经验')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('5', '5年工作经验')
cc['工作岗位要求经验'] = cc['工作岗位要求经验'].replace('10', '10年以上工作经验')

cc['工作经验'] = cc['工作经验'].replace('无经验', '经验不限')
cc

工作id        工作岗位名称  工作岗位最低工资 工作岗位要求经验  \
0     1.631113e+18          会计实习生            3500.0         经验不限   
1     1.631113e+18      技术服务工程师            5000.0         经验不限   
2     1.629825e+18  大数据分析师（BI）            4500.0         经验不限   
3     1.613440e+18  自然语言处理工程师            2000.0         经验不限   
4     1.613440e+18          爬虫工程师            2000.0         经验不限   
...            ...                 ...               ...              ...   
1569  1.463032e+18      前端开发工程师            8000.0         经验不限   
1570  1.374181e+18      软件测试工程师            6000.0      1年工作经验   
1571  1.374177e+18      新媒体运营助理            5000.0      1年工作经验   
1572  1.374178e+18        数据产品经理          150000.0      3年工作经验   
1573  1.374197e+18      数据分析实习生             100.0         经验不限   

                                        岗位技能               人才id  \
0                                           None  1635610357617786880   
1                                           None  1634019092572798976   
2                                           None  1632947816114552832   
3                               算法研究深度学习  1632286586903330816   
4                         Python爬虫框架数据采集  1579833809761861632   
...                                          ...                  ...   
1569  数据查询数据计算数据挖掘数据分析数据可视化  7535876391727459860   
1570                        系统测试软件系统搭建  7535876400317394452   
1571                                    产品运营  7535876404612361748   
1572  数据探索数据可视化项目管理需求分析产品设计  7535876413202296340   
1573                    数据探索数据查询数据建模  7535876417497263636   

                                          应聘工作岗位  最低薪资期望  \
0         ["数据分析师","数据挖掘工程师","算法工程师"]          3000   
1                                       ["数据分析师"]          4000   
2     ["数据分析师","数据挖掘工程师","机器学习工程师"]          6000   
3     ["数据分析师","数据挖掘工程师","机器学习工程师"]          4000   
4                      ["数据分析师","数据挖掘工程师"]          4000   
...                                                ...           ...   
1569                   ["数据分析师","数据挖掘工程师"]          4000   
1570                   ["数据分析师","数据挖掘工程师"]          4000   
1571                   ["数据分析师","数据挖掘工程师"]          4000   
1572                   ["数据分析师","数据挖掘工程师"]          4000   
1573                   ["数据分析师","数据挖掘工程师"]          4000   

         工作经验 个人技能  
0        经验不限     None  
1        经验不限     None  
2     1年工作经验     None  
3        经验不限     None  
4        经验不限     None  
...           ...      ...  
1569     经验不限     None  
1570     经验不限     None  
1571     经验不限     None  
1572     经验不限     None  
1573     经验不限     None  

[1574 rows x 10 columns]

##### 将岗位内没有特殊说明技能要求的统一为空值

In [10]:
cc['岗位技能'] = cc['岗位技能'].replace('None', np.nan)
cc['个人技能'] = cc['个人技能'].replace('None', np.nan)
cc.head(3)

工作id        工作岗位名称  工作岗位最低工资 工作岗位要求经验  \
0  1.631113e+18          会计实习生            3500.0         经验不限   
1  1.631113e+18      技术服务工程师            5000.0         经验不限   
2  1.629825e+18  大数据分析师（BI）            4500.0         经验不限   

  岗位技能               人才id  \
0      NaN  1635610357617786880   
1      NaN  1634019092572798976   
2      NaN  1632947816114552832   

                                       应聘工作岗位  最低薪资期望  \
0      ["数据分析师","数据挖掘工程师","算法工程师"]          3000   
1                                    ["数据分析师"]          4000   
2  ["数据分析师","数据挖掘工程师","机器学习工程师"]          6000   

      工作经验 个人技能  
0     经验不限      NaN  
1     经验不限      NaN  
2  1年工作经验      NaN

##### 清洗应聘多种工作岗位的数据，["数据分析师","数据挖掘工程师","机器学习工程师"]中的中括号，双引号，逗号等

In [11]:
cc['应聘工作岗位'] = cc['应聘工作岗位'].apply(lambda x:x.replace('[',''))
cc['应聘工作岗位'] = cc['应聘工作岗位'].apply(lambda x:x.replace(']',''))
cc['应聘工作岗位'] = cc['应聘工作岗位'].apply(lambda x:x.replace('"',''))
cc['应聘工作岗位'] = cc['应聘工作岗位'].apply(lambda x:x.replace(',',''))
cc

工作id        工作岗位名称  工作岗位最低工资 工作岗位要求经验  \
0     1.631113e+18          会计实习生            3500.0         经验不限   
1     1.631113e+18      技术服务工程师            5000.0         经验不限   
2     1.629825e+18  大数据分析师（BI）            4500.0         经验不限   
3     1.613440e+18  自然语言处理工程师            2000.0         经验不限   
4     1.613440e+18          爬虫工程师            2000.0         经验不限   
...            ...                 ...               ...              ...   
1569  1.463032e+18      前端开发工程师            8000.0         经验不限   
1570  1.374181e+18      软件测试工程师            6000.0      1年工作经验   
1571  1.374177e+18      新媒体运营助理            5000.0      1年工作经验   
1572  1.374178e+18        数据产品经理          150000.0      3年工作经验   
1573  1.374197e+18      数据分析实习生             100.0         经验不限   

                                        岗位技能               人才id  \
0                                            NaN  1635610357617786880   
1                                            NaN  1634019092572798976   
2                                            NaN  1632947816114552832   
3                               算法研究深度学习  1632286586903330816   
4                         Python爬虫框架数据采集  1579833809761861632   
...                                          ...                  ...   
1569  数据查询数据计算数据挖掘数据分析数据可视化  7535876391727459860   
1570                        系统测试软件系统搭建  7535876400317394452   
1571                                    产品运营  7535876404612361748   
1572  数据探索数据可视化项目管理需求分析产品设计  7535876413202296340   
1573                    数据探索数据查询数据建模  7535876417497263636   

                                应聘工作岗位  最低薪资期望     工作经验  \
0         数据分析师数据挖掘工程师算法工程师          3000     经验不限   
1                                 数据分析师          4000     经验不限   
2     数据分析师数据挖掘工程师机器学习工程师          6000  1年工作经验   
3     数据分析师数据挖掘工程师机器学习工程师          4000     经验不限   
4                   数据分析师数据挖掘工程师          4000     经验不限   
...                                      ...           ...          ...   
1569                数据分析师数据挖掘工程师          4000     经验不限   
1570                数据分析师数据挖掘工程师          4000     经验不限   
1571                数据分析师数据挖掘工程师          4000     经验不限   
1572                数据分析师数据挖掘工程师          4000     经验不限   
1573                数据分析师数据挖掘工程师          4000     经验不限   

     个人技能  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
...       ...  
1569      NaN  
1570      NaN  
1571      NaN  
1572      NaN  
1573      NaN  

[1574 rows x 10 columns]

In [12]:
cc.to_csv('F:/amyself/c题爬虫/第三题/构建岗位匹配度模型.csv',encoding='utf8')

In [13]:
cc["xdate"] = cc["工作岗位名称"].map(str) + cc["工作岗位最低工资"].map(str) + cc["工作岗位要求经验"].map(str) + cc["岗位技能"].map(str)  
cc["ydate"] = cc["应聘工作岗位"].map(str) + cc["最低薪资期望"].map(str) + cc["工作经验"].map(str) + cc["个人技能"].map(str)  
cc

工作id        工作岗位名称  工作岗位最低工资 工作岗位要求经验  \
0     1.631113e+18          会计实习生            3500.0         经验不限   
1     1.631113e+18      技术服务工程师            5000.0         经验不限   
2     1.629825e+18  大数据分析师（BI）            4500.0         经验不限   
3     1.613440e+18  自然语言处理工程师            2000.0         经验不限   
4     1.613440e+18          爬虫工程师            2000.0         经验不限   
...            ...                 ...               ...              ...   
1569  1.463032e+18      前端开发工程师            8000.0         经验不限   
1570  1.374181e+18      软件测试工程师            6000.0      1年工作经验   
1571  1.374177e+18      新媒体运营助理            5000.0      1年工作经验   
1572  1.374178e+18        数据产品经理          150000.0      3年工作经验   
1573  1.374197e+18      数据分析实习生             100.0         经验不限   

                                        岗位技能               人才id  \
0                                            NaN  1635610357617786880   
1                                            NaN  1634019092572798976   
2                                            NaN  1632947816114552832   
3                               算法研究深度学习  1632286586903330816   
4                         Python爬虫框架数据采集  1579833809761861632   
...                                          ...                  ...   
1569  数据查询数据计算数据挖掘数据分析数据可视化  7535876391727459860   
1570                        系统测试软件系统搭建  7535876400317394452   
1571                                    产品运营  7535876404612361748   
1572  数据探索数据可视化项目管理需求分析产品设计  7535876413202296340   
1573                    数据探索数据查询数据建模  7535876417497263636   

                                应聘工作岗位  最低薪资期望     工作经验  \
0         数据分析师数据挖掘工程师算法工程师          3000     经验不限   
1                                 数据分析师          4000     经验不限   
2     数据分析师数据挖掘工程师机器学习工程师          6000  1年工作经验   
3     数据分析师数据挖掘工程师机器学习工程师          4000     经验不限   
4                   数据分析师数据挖掘工程师          4000     经验不限   
...                                      ...           ...          ...   
1569                数据分析师数据挖掘工程师          4000     经验不限   
1570                数据分析师数据挖掘工程师          4000     经验不限   
1571                数据分析师数据挖掘工程师          4000     经验不限   
1572                数据分析师数据挖掘工程师          4000     经验不限   
1573                数据分析师数据挖掘工程师          4000     经验不限   

     个人技能  \
0         NaN   
1         NaN   
2         NaN   
3         NaN   
4         NaN   
...       ...   
1569      NaN   
1570      NaN   
1571      NaN   
1572      NaN   
1573      NaN   

                                                                                xdate  \
0                           会计实习生3500.0经验不限nan                                 
1                       技术服务工程师5000.0经验不限nan                                 
2                   大数据分析师（BI）4500.0经验不限nan                                 
3      自然语言处理工程师2000.0经验不限算法研究深度学习                                 
4        爬虫工程师2000.0经验不限Python爬虫框架数据采集                                 
...                                                 ...                                 
1569  前端开发工程师8000.0经验不限数据查询数据计算数据挖掘数据分析数据可视化......      
1570  软件测试工程师6000.01年工作经验系统测试软件系统搭建......                         
1571            新媒体运营助理5000.01年工作经验产品运营                                 
1572  数据产品经理150000.03年工作经验数据探索数据可视化项目管理需求分析产品设计......   
1573  数据分析实习生100.0经验不限数据探索数据查询数据建模......                         

                                                               ydate  
0     数据分析师数据挖掘工程师算法工程师3000经验不限nan               
1                             数据分析师4000经验不限nan               
2     数据分析师数据挖掘工程师机器学习工程师60001年工作经验nan......  
3     数据分析师数据挖掘工程师机器学习工程师4000经验不限nan......     
4               数据分析师数据挖掘工程师4000经验不限nan               
...                                                 ...               
1569            数据分析师数据挖掘工程师4000经验不限nan               
1570            数据分析师数据挖掘工程师4000经验不限nan               
1571            数据分析师数据挖掘工程师4000经验不限nan               
1572            数据分析师数据挖掘工程师4000经验不限nan               
1573            数据分析师数据挖

In [14]:
hh = ['工作id','xdate','人才id','ydate']
hha = cc[hh]
hha

工作id  \
0     1.631113e+18   
1     1.631113e+18   
2     1.629825e+18   
3     1.613440e+18   
4     1.613440e+18   
...            ...   
1569  1.463032e+18   
1570  1.374181e+18   
1571  1.374177e+18   
1572  1.374178e+18   
1573  1.374197e+18   

                                                                                xdate  \
0                           会计实习生3500.0经验不限nan                                 
1                       技术服务工程师5000.0经验不限nan                                 
2                   大数据分析师（BI）4500.0经验不限nan                                 
3      自然语言处理工程师2000.0经验不限算法研究深度学习                                 
4        爬虫工程师2000.0经验不限Python爬虫框架数据采集                                 
...                                                 ...                                 
1569  前端开发工程师8000.0经验不限数据查询数据计算数据挖掘数据分析数据可视化......      
1570  软件测试工程师6000.01年工作经验系统测试软件系统搭建......                         
1571            新媒体运营助理5000.01年工作经验产品运营                                 
1572  数据产品经理150000.03年工作经验数据探索数据可视化项目管理需求分析产品设计......   
1573  数据分析实习生100.0经验不限数据探索数据查询数据建模......                         

                   人才id  \
0     1635610357617786880   
1     1634019092572798976   
2     1632947816114552832   
3     1632286586903330816   
4     1579833809761861632   
...                   ...   
1569  7535876391727459860   
1570  7535876400317394452   
1571  7535876404612361748   
1572  7535876413202296340   
1573  7535876417497263636   

                                                               ydate  
0     数据分析师数据挖掘工程师算法工程师3000经验不限nan               
1                             数据分析师4000经验不限nan               
2     数据分析师数据挖掘工程师机器学习工程师60001年工作经验nan......  
3     数据分析师数据挖掘工程师机器学习工程师4000经验不限nan......     
4               数据分析师数据挖掘工程师4000经验不限nan               
...                                                 ...               
1569            数据分析师数据挖掘工程师4000经验不限nan               
1570            数据分析师数据挖掘工程师4000经验不限nan               
1571            数据分析师数据挖掘工程师4000经验不限nan               
1572            数据分析师数据挖掘工程师4000经验不限nan               
1573            数据分析师数据挖掘工程师4000经验不限nan               

[1574 rows x 4 columns]

##### 将“工作岗位名称”和“应聘工作岗位”一些尾部的nan取代为空值

In [15]:
hha['xdate'] = hha['xdate'].apply(lambda x:x.replace('nan',''))
hha['ydate'] = hha['ydate'].apply(lambda x:x.replace('nan',''))
hha # 报红是提醒不是报错，是说有一部分数据里面没有空值

C:\Users\Administrator\AppData\Local\Temp\ipykernel_17180\869358114.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hha['xdate'] = hha['xdate'].apply(lambda x:x.replace('nan',''))
C:\Users\Administrator\AppData\Local\Temp\ipykernel_17180\869358114.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hha['ydate'] = hha['ydate'].apply(lambda x:x.replace('nan',''))


工作id  \
0     1.631113e+18   
1     1.631113e+18   
2     1.629825e+18   
3     1.613440e+18   
4     1.613440e+18   
...            ...   
1569  1.463032e+18   
1570  1.374181e+18   
1571  1.374177e+18   
1572  1.374178e+18   
1573  1.374197e+18   

                                                                                xdate  \
0                              会计实习生3500.0经验不限                                 
1                          技术服务工程师5000.0经验不限                                 
2                      大数据分析师（BI）4500.0经验不限                                 
3      自然语言处理工程师2000.0经验不限算法研究深度学习                                 
4        爬虫工程师2000.0经验不限Python爬虫框架数据采集                                 
...                                                 ...                                 
1569  前端开发工程师8000.0经验不限数据查询数据计算数据挖掘数据分析数据可视化......      
1570  软件测试工程师6000.01年工作经验系统测试软件系统搭建......                         
1571            新媒体运营助理5000.01年工作经验产品运营                                 
1572  数据产品经理150000.03年工作经验数据探索数据可视化项目管理需求分析产品设计......   
1573  数据分析实习生100.0经验不限数据探索数据查询数据建模......                         

                   人才id  \
0     1635610357617786880   
1     1634019092572798976   
2     1632947816114552832   
3     1632286586903330816   
4     1579833809761861632   
...                   ...   
1569  7535876391727459860   
1570  7535876400317394452   
1571  7535876404612361748   
1572  7535876413202296340   
1573  7535876417497263636   

                                                            ydate  
0        数据分析师数据挖掘工程师算法工程师3000经验不限            
1                                数据分析师4000经验不限            
2     数据分析师数据挖掘工程师机器学习工程师60001年工作经验......  
3     数据分析师数据挖掘工程师机器学习工程师4000经验不限......     
4                  数据分析师数据挖掘工程师4000经验不限            
...                                                 ...            
1569               数据分析师数据挖掘工程师4000经验不限            
1570               数据分析师数据挖掘工程师4000经验不限            
1571               数据分析师数据挖掘工程师4000经验不限            
1572               数据分析师数据挖掘工程师4000经验不限            
1573               数据分析师数据挖掘工程师4000经验不限            

[1574 rows x 4 columns]

In [16]:
import csv,shutil
hha.to_csv('F:/amyself/c题爬虫/第三题/haha.csv',encoding='utf8')

a=open('F:/amyself/c题爬虫/第三题/hha.csv','r',encoding='utf8')
reader = csv.reader(a)
 
with open('F:/amyself/c题爬虫/第三题/hha.txt','w',encoding='utf8') as f:
    
    for i in reader:
        for x in i:
            f.write(x)
            f.write('\t')
        f.write('\n')
a.close()


In [17]:
import jieba

# 添加自定义词典
jieba.load_userdict("F:/amyself/c题爬虫/第三题/hha.txt")
# 分词测试
seg_words = jieba.lcut("高级大数据开发工程师 Hadoop Hive20000.0经验不限")
print(seg_words)

['高级', '大', '数据', '开发', '工程师', ' ', 'Hadoop', ' ', 'Hive20000.0', '经验', '不', '限']


In [18]:
hha.head(2)

工作id                         xdate               人才id  \
0  1.631113e+18      会计实习生3500.0经验不限  1635610357617786880   
1  1.631113e+18  技术服务工程师5000.0经验不限  1634019092572798976   

                                            ydate  
0  数据分析师数据挖掘工程师算法工程师3000经验不限  
1                          数据分析师4000经验不限

In [19]:
hha["xdate"] = hha["xdate"].apply(lambda x: jieba.lcut(x.strip("\r\t ").replace("***","*")))
hha["ydate"] = hha["ydate"].apply(lambda x: jieba.lcut(x.strip("\r\t ").replace("***","*")))
hha.head()

C:\Users\Administrator\AppData\Local\Temp\ipykernel_17180\1998705958.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hha["xdate"] = hha["xdate"].apply(lambda x: jieba.lcut(x.strip("\r\t ").replace("***","*")))
C:\Users\Administrator\AppData\Local\Temp\ipykernel_17180\1998705958.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hha["ydate"] = hha["ydate"].apply(lambda x: jieba.lcut(x.strip("\r\t ").replace("***","*")))


工作id  \
0  1.631113e+18   
1  1.631113e+18   
2  1.629825e+18   
3  1.613440e+18   
4  1.613440e+18   

                                                                  xdate  \
0               [会计, 实习生, 3500.0, 经验, 不, 限]                      
1         [技术, 服务, 工程师, 5000.0, 经验, 不, 限]                      
2  [大, 数据, 分析师, （, BI, ）, 4500.0, 经验, 不, 限]......             
3  [自然语言, 处理, 工程师, 2000.0, 经验, 不, 限, 算法, 研究, 深度, ...   
4  [爬虫, 工程师, 2000.0, 经验, 不限, Python, 爬虫, 框架, 数据, ...       

                人才id  \
0  1635610357617786880   
1  1634019092572798976   
2  1632947816114552832   
3  1632286586903330816   
4  1579833809761861632   

                                                                     ydate  
0  [数据, 分析师, 数据挖掘, 工程师, 算法, 工程师, 3000, 经验, 不, 限]....   
1                 [数据, 分析师, 4000, 经验, 不, 限]                        
2  [数据, 分析师, 数据挖掘, 工程师, 机器, 学习, 工程师, 60001, 年, 工作...  
3  [数据, 分析师, 数据挖掘, 工程师, 机器, 学习, 工程师, 4000, 经验, 不,...  
4  [数据, 分析师, 数据挖掘, 工程师, 4000, 经验, 不, 限]......

##### 添加一列“label”存放匹配度，默认值为0

In [20]:
hha['label'] = 0
hha

C:\Users\Administrator\AppData\Local\Temp\ipykernel_17180\3645331908.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hha['label'] = 0


工作id  \
0     1.631113e+18   
1     1.631113e+18   
2     1.629825e+18   
3     1.613440e+18   
4     1.613440e+18   
...            ...   
1569  1.463032e+18   
1570  1.374181e+18   
1571  1.374177e+18   
1572  1.374178e+18   
1573  1.374197e+18   

                                                                      xdate  \
0                  [会计, 实习生, 3500.0, 经验, 不, 限]                       
1            [技术, 服务, 工程师, 5000.0, 经验, 不, 限]                       
2     [大, 数据, 分析师, （, BI, ）, 4500.0, 经验, 不, 限]......              
3     [自然语言, 处理, 工程师, 2000.0, 经验, 不, 限, 算法, 研究, 深度, ...    
4     [爬虫, 工程师, 2000.0, 经验, 不限, Python, 爬虫, 框架, 数据, ...        
...                                                                ...        
1569  [前端开发, 工程师, 8000.0, 经验, 不, 限, 数据, 查询, 数据, 计算, ...    
1570  [软件测试, 工程师, 6000.01, 年, 工作, 经验, 系统, 测试, 软件系统, ...   
1571  [新, 媒体, 运营, 助理, 5000.01, 年, 工作, 经验, 产品, 运营]......       
1572  [数据, 产品, 经理, 150000.03, 年, 工作, 经验, 数据, 探索, 数据,...      
1573  [数据分析, 实习生, 100.0, 经验, 不, 限, 数据, 探索, 数据, 查询, 数...   

                   人才id  \
0     1635610357617786880   
1     1634019092572798976   
2     1632947816114552832   
3     1632286586903330816   
4     1579833809761861632   
...                   ...   
1569  7535876391727459860   
1570  7535876400317394452   
1571  7535876404612361748   
1572  7535876413202296340   
1573  7535876417497263636   

                                                                        ydate  \
0     [数据, 分析师, 数据挖掘, 工程师, 算法, 工程师, 3000, 经验, 不, 限]....    
1                    [数据, 分析师, 4000, 经验, 不, 限]                         
2     [数据, 分析师, 数据挖掘, 工程师, 机器, 学习, 工程师, 60001, 年, 工作...   
3     [数据, 分析师, 数据挖掘, 工程师, 机器, 学习, 工程师, 4000, 经验, 不,...   
4     [数据, 分析师, 数据挖掘, 工程师, 4000, 经验, 不, 限]......                
...                                                          ...                
1569  [数据, 分析师, 数据挖掘, 工程师, 4000, 经验, 不, 限]......                
1570  [数据, 分析师, 数据挖掘, 工程师, 4000, 经验, 不, 限]......                
1571  [数据, 分析师, 数据挖掘, 工程师, 4000, 经验, 不, 限]......                
1572  [数据, 分析师, 数据挖掘, 工程师, 4000, 经验, 不, 限]......                
1573  [数据, 分析师, 数据挖掘, 工程师, 4000, 经验, 不, 限]......                

      label  
0         0  
1         0  
2         0  
3         0  
4         0  
...     ...  
1569      0  
1570      0  
1571      0  
1572      0  
1573      0  

[1574 rows x 5 columns]

##### 运用结巴库分词后统计每个分词的数量

In [21]:
from collections import Counter

c = Counter()
sent_data = hha["xdate"].values + hha["xdate"].values
for d in sent_data:
    c.update(d)
word_counts = sorted(dict(c).items(), key=lambda x: x[1], reverse=True)

print(word_counts[:10]) # 获取前10个分词数量进行展示

[('经验', 3148), ('数据', 2408), ('工作', 2210), ('年', 2208), ('分析师', 1284), ('不', 922), ('限', 922), ('产品', 720), ('经理', 634), ('工程师', 608)]


##### 获取分词相对应的PAD、UNK值

In [22]:
vocab_words = ["<PAD>", "<UNK>"]
for w, c in word_counts:
    vocab_words.append(w)

vocab2id = {w: i for i, w in enumerate(vocab_words)}
id2vocab = {i: w for i, w in enumerate(vocab_words)}

print("vocab size: ", len(vocab2id))
print(list(vocab2id.items())[:5])
print(list(id2vocab.items())[:5])

vocab size:  800
[('<PAD>', 0), ('<UNK>', 1), ('经验', 2), ('数据', 3), ('工作', 4)]
[(0, '<PAD>'), (1, '<UNK>'), (2, '经验'), (3, '数据'), (4, '工作')]


##### 将训练分词结果存储进vocab.txt文本文件中

In [23]:
with open("F:/amyself/c题爬虫/第三题/vocab.txt", "w", encoding="utf8") as f:
    for w, i in vocab2id.items():
        f.write(w+"\n")

In [24]:
def sent2index(vocab2id, words):
    return [vocab2id[w] if w in vocab2id else vocab2id["<UNK>"] for w in words]

hha["xdate"] = hha["xdate"].apply(lambda x: sent2index(vocab2id, x))
hha["ydate"] = hha["ydate"].apply(lambda x: sent2index(vocab2id, x))

hha.head(5)

C:\Users\Administrator\AppData\Local\Temp\ipykernel_17180\926855422.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hha["xdate"] = hha["xdate"].apply(lambda x: sent2index(vocab2id, x))
C:\Users\Administrator\AppData\Local\Temp\ipykernel_17180\926855422.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hha["ydate"] = hha["ydate"].apply(lambda x: sent2index(vocab2id, x))


工作id                              xdate               人才id  \
0  1.631113e+18                 [1, 1, 1, 1, 1, 1]  1635610357617786880   
1  1.631113e+18              [1, 1, 1, 1, 1, 1, 1]  1634019092572798976   
2  1.629825e+18     [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]  1632947816114552832   
3  1.613440e+18  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]  1632286586903330816   
4  1.613440e+18     [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]  1579833809761861632   

                               ydate  label  
0     [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]      0  
1                 [1, 1, 1, 1, 1, 1]      0  
2  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]      0  
3  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]      0  
4           [1, 1, 1, 1, 1, 1, 1, 1]      0

In [25]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import backend as K

print(tf.__version__)

class BaseTextCNN(keras.Model):
    def __init__(self, filters, kernel_sizes, output_dim, name):
        super(BaseTextCNN, self).__init__(name=name)
        self.kernel_sizes = kernel_sizes
        self.conv_layers = []
        self.max_poolings = []
        for kernel_size in kernel_sizes:
            self.conv_layers.append(
                keras.layers.Conv1D(filters=filters, kernel_size=kernel_size, 
                                activation='relu', padding="same")
            )
            self.max_poolings.append(keras.layers.GlobalMaxPool1D())
        self.concatenate = keras.layers.Concatenate()
        self.dense = keras.layers.Dense(output_dim, activation='tanh')
        
    def call(self, inputs):
        convs = []
        for i in range(len(self.kernel_sizes)):
            x = self.conv_layers[i](inputs)
            x = self.max_poolings[i](x)
            convs.append(x)
        x = self.concatenate(convs)
        output = self.dense(x)
        return output

2.3.0


In [26]:
max_len = 15
vocab_size = len(vocab2id)
embedding_size = 128
filters = 200
kernel_sizes = [3,4,5]
output_dim = 100

In [27]:
from tensorflow.keras.layers import Input,Embedding,Dot,Dense
from tensorflow.keras.models import Model

In [28]:
from tensorflow.keras.layers import Dense,Conv1D,GlobalMaxPool1D,concatenate
from tensorflow.keras.layers import Bidirectional,LSTM
from tensorflow.keras.layers import Attention

In [29]:
# EMBEDDING_SIZE = 100
hidden_size= 64
attention_size = 50
# hidden_size=64
class_nums=2

In [30]:
from tensorflow.keras import  initializers,regularizers,constraints
from  tensorflow.keras.layers import Layer
import tensorflow as tf

class Attention(Layer):
    def __init__(self,
                 W_regularizer=None,
                 b_regularizer=None,
                 W_constraint=None,
                 b_constraint=None,
                 bias=True,
                 **kwargs
                 ):
        """
        Keras Layer that implements an Attention mechanism for temporal data.
        Supports Masking.
        Follows the work of Raffel et al. [https://arxiv.org/abs/1512.08756]
        # Input shape
            3D tensor with shape: `(samples, steps, features)`.
        # Output shape
            2D tensor with shape: `(samples, features)`.
        :param kwargs:
        Just put it on top of an RNN Layer (GRU/LSTM/SimpleRNN) with return_sequences=True.
        The dimensions are inferred based on the output shape of the RNN.
        Example:
            # 1
            model.add(LSTM(64, return_sequences=True))
            model.add(Attention())
            # next add a Dense layer (for classification/regression) or whatever...
            # 2
            hidden = LSTM(64, return_sequences=True)(words)
            sentence = Attention()(hidden)
            # next add a Dense layer (for classification/regression) or whatever...
        """
        super(Attention, self).__init__()
        self.bias = bias
        self.init = initializers.get('glorot_uniform')

    def build(self, input_shape):
        '''
        :param input_shape:
        :return:
        '''
        self.output_dim = input_shape[-1]
        self.W = self.add_weight(
                                 name='{}_W'.format(self.name),
                                 shape=(input_shape[2], 1),
                                 initializer=self.init,
                                 trainable=True
                                 )
        if self.bias:
            self.b = self.add_weight(
                                     name='{}_b'.format(self.name),
                                     shape=(input_shape[1], 1),
                                     initializer='zero',
                                     trainable=True
                                     )
        else:
            self.b = None

        self.built = True

    def compute_mask(self, inputs, mask=None):
        return None

    def call(self, inputs, mask=None):
        # (N, step, d), (d, 1)  ==>   (N, step, 1)
        e = tf.matmul(inputs, self.W, )
        if self.bias:
            e += self.b
        e = tf.tanh(e)
        a = tf.nn.softmax(e, axis=1)
        # (N, step, d) (N, step, 1) ====> (N, step, d)
        c = inputs*a
        # (N, d)
        c = tf.reduce_sum(c, axis=1)
        return c

    def get_config(self):
        return {'units': self.output_dim}

In [31]:
input1=Input(name='sent1',shape=(max_len,))
input2=Input(name='sent2',shape=(max_len,))
embedding=Embedding(vocab_size,embedding_size)
sent1_embed=embedding(input1)
sent2_embed=embedding(input2)
b1=Bidirectional(LSTM(hidden_size, dropout=0.2, return_sequences=True))(sent1_embed)
b2=Bidirectional(LSTM(hidden_size, dropout=0.2, return_sequences=True))(sent2_embed)
output_sent1= Attention()(b1)
output_sent2= Attention()(b2)
cosine_output=Dot(axes=[1,1],normalize=True)([output_sent1,output_sent2])
outputs=Dense(1,activation='linear',name="output")(cosine_output)
model=Model(inputs=[input1,input2],outputs=outputs)

In [32]:
model.summary()

Model: "functional_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
sent1 (InputLayer)              [(None, 15)]         0                                            
__________________________________________________________________________________________________
sent2 (InputLayer)              [(None, 15)]         0                                            
__________________________________________________________________________________________________
embedding (Embedding)           (None, 15, 128)      102400      sent1[0][0]                      
                                                                 sent2[0][0]                      
__________________________________________________________________________________________________
bidirectional (Bidirectional)   (None, 15, 128)      98816       embedding[0][0]       

In [33]:
import numpy as np
from tensorflow.keras.preprocessing.sequence import pad_sequences

def batch_generator(all_data, batch_size, maxlen, shuffle=True):
    """
    :param all_data : all_data整个数据集，包含输入和输出标签
    :param batch_size: batch_size表示每个batch的大小
    :param shuffle: 是否打乱顺序
    :return:
    """
    # 输入all_datas的每一项必须是numpy数组，保证后面能按p所示取值
    all_data = [np.array(d) for d in all_data]
    # 获取样本大小
    data_size = all_data[0].shape[0]

    if shuffle:
        # 随机生成打乱的索引
        p = np.random.permutation(data_size)
        # 重新组织数据
        all_data = [d[p] for d in all_data]
        
    batch_count = 0
    while True:
        # 数据一轮循环(epoch)完成，打乱一次顺序
        if batch_count * batch_size + batch_size > data_size:
            batch_count = 0
            if shuffle:
                p = np.random.permutation(data_size)
                all_data = [d[p] for d in all_data]
        start = batch_count * batch_size
        end = start + batch_size
        batch_count += 1
        batch_data = [d[start: end] for d in all_data]
        batch_sent1, batch_sent2, batch_label = batch_data
        
        batch_sent1_pad = pad_sequences(batch_sent1, maxlen=max_len, padding='post')
        batch_sent2_pad = pad_sequences(batch_sent2, maxlen=max_len, padding='post')
        
        yield [batch_sent1_pad, batch_sent2_pad], batch_label

In [34]:
sent1_datas = hha["xdate"].values.tolist()
sent2_datas = hha["xdate"].values.tolist()
labels = hha["label"].values.tolist()

In [35]:
# 划分训练 测试数据集
count = len(labels)
idx1, idx2 = int(count*0.8), int(count*0.9)
sent1_train, sent2_train = sent1_datas[:idx1], sent2_datas[:idx1]
sent1_val, sent2_val = sent1_datas[idx1:idx2], sent2_datas[idx1:idx2]
sent1_test, sent2_test = sent1_datas[idx2:], sent2_datas[idx2:]

train_labels, val_labels, test_labels = labels[:idx1], labels[idx1:idx2], labels[idx2:]

print("train data: ", len(sent1_train), len(sent2_train), len(train_labels))
print("val data: ", len(sent1_val), len(sent2_val), len(val_labels))
print("test data: ", len(sent1_test), len(sent2_test), len(test_labels))

train data:  1259 1259 1259
val data:  157 157 157
test data:  158 158 158


In [36]:
# batch数据的生成器
batch_size = 64
maxlen = 15
batch_count = int(len(train_labels) / batch_size)
batch_gen_train = batch_generator([sent1_train, sent2_train, train_labels], batch_size, max_len)
batch_gen_val = batch_generator([sent1_val, sent2_val, val_labels], batch_size, max_len)
batch_gen_test = batch_generator([sent1_test, sent2_test, test_labels], batch_size, max_len)

In [37]:
epochs = 10

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# 训练模型
model.fit(batch_gen_train, 
                    verbose=1, 
                    validation_data=batch_gen_val,
                    validation_steps=100,
                    steps_per_epoch=batch_count, 
                    epochs=2)

# 预测模型
score, acc = model.evaluate(batch_gen_test, steps=50, 
                                      max_queue_size=10, 
                                      use_multiprocessing=False)
print('score:', score, 'accuracy:', acc)

# 保存训练好的模型
# model.save("output/cnndssm_semantic_match.h5")
# model.save_weights("output/match_model_weight.h5")

Epoch 1/2
19/19 [==============================] - 3s 145ms/step - loss: 8.7103e-06 - accuracy: 1.0000 - val_loss: 0.0000e+00 - val_accuracy: 1.0000
Epoch 2/2
50/50 [==============================] - 1s 12ms/step - loss: 0.0000e+00 - accuracy: 1.0000
score: 0.0 accuracy: 1.0


##### 随机提取两个岗位信息做分词匹配测试

In [38]:
import math
def sent2index(vocab2id, words):
    return [vocab2id[w] if w in vocab2id else vocab2id["<UNK>"] for w in words]

sent1 = "高级大数据开发工程师 Hadoop Hive20000.0经验不限"
sent2 = "项目经理产品经理资质管理员电商大数据研究员60001年工作经验"

sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))

sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')

# model.load_weights("output/match_model_weight.h5")

preds = model.predict([sent1_pad, sent2_pad])

print("sent1: %s" % sent1)
print("sent2: %s" % sent2)
print("score: %s" % math.fabs(preds[0]))

sent1: 高级大数据开发工程师 Hadoop Hive20000.0经验不限
sent2: 项目经理产品经理资质管理员电商大数据研究员60001年工作经验
score: 0.36315152049064636


#### 测试结果显示："高级大数据开发工程师 Hadoop Hive20000.0经验不限"与"项目经理产品经理资质管理员电商大数据研究员60001年工作经"两个字符串之间的匹配度为"0.36315152049064636"，即该求职者所具有的技能经验和对岗位的要求工资等信息与该岗位需要具备的内容相关性只有“0.36315152049064636”-相关性有但不多

#### 根据题目四需要对匹配度进行非‘0’，即‘1’，的转化筛选，会对所有的匹配值求平均值，匹配值高于或等于平均匹配度设置为‘1’，低于平均匹配度设置为‘0’。

In [41]:
haha = pd.read_csv('F:/amyself/c题爬虫/第三题/haha.csv',encoding='utf-8')
haha = pd.DataFrame(haha)
haha['label'] = 0
haha_x = haha['xdate']
haha_y = haha['ydate']
haha_x1 = []
haha_y1 = []
for i in haha_x:
    haha_x1.append(i)
for j in haha_y:
    haha_y1.append(j)

##### 由于在该列表情况下并不是每一个求职者固定对应到每个岗位，所以将求职者与每一个岗位的关键字进行匹配训练，取平均值作为该岗位面对求职者的匹配度

##### 获取前50个应聘者的数据进行匹配度训练，每个应聘者训练1574个岗位数据，得出结果如下

In [40]:
haha_label = []
# haha_label_xy = []
for sent1 in haha['xdate'][0:50]:
    for sent2 in haha['ydate']:
        sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
        sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
        sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
        sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
        preds = model.predict([sent1_pad, sent2_pad])
        haha_label.append(math.fabs(preds[0]))
#     haha_mean = sum(haha_label_xy)/len(haha_label_xy)
#     haha_label.append(haha_label_xy)
haha_label

[0.7608605027198792,
 0.8429649472236633,
 0.7389992475509644,
 0.7329651713371277,
 0.8201602697372437,
 0.8022873401641846,
 0.7999355792999268,
 0.8443192839622498,
 0.8201602697372437,
 0.8443192839622498,
 0.6654328107833862,
 0.8461789488792419,
 0.7608605027198792,
 0.7608605027198792,
 0.8443192839622498,
 0.8355852365493774,
 0.8429649472236633,
 0.8201602697372437,
 0.8429649472236633,
 0.7535282373428345,
 0.7389992475509644,
 0.7420763373374939,
 0.8070203065872192,
 0.7554445266723633,
 0.8429649472236633,
 0.8201602697372437,
 0.6041468381881714,
 0.836408257484436,
 0.8500115871429443,
 0.41945263743400574,
 0.5857394933700562,
 0.8429649472236633,
 0.7373352646827698,
 0.38211432099342346,
 0.2986070513725281,
 0.42405104637145996,
 0.38708069920539856,
 0.8429649472236633,
 0.7942347526550293,
 0.8229609131813049,
 0.7329651713371277,
 0.329683780670166,
 0.7999355792999268,
 0.8429649472236633,
 0.798346221446991,
 0.8429649472236633,
 0.8070203065872192,
 0.734932065

##### 查看整体，50个应聘者，每个应聘者匹配1574个岗位数据，共计78700条数据

In [45]:
len(haha_label)

78700

### 将每个求职者的去匹配到每一个岗位，根据前面分析，我们得到1574条岗位数据，10908个求职者，在求职者中因为空值数据，错误数据，异常数据等缘故，在整合表格的过程中选取到的求职者数据也是1574条，此时每个求职者对应1574个岗位去做分析，共有1574个求职者去遍历建立模型，可以总共得到“2,477,476”条数据

### 由于jupyter notebook中无法做多进程，以下多进程调用模型去做匹配度分析的代码是在PyCharm中进行的

### 由于在内存内核中无法一次性对两百万条数据进行访问、处理和存储，因此使用数据划分、将每一百个求职者，即对应“157,400”条数据做为一个进程去做存储与分析，共计16个进程同时并发去访问数据源，达到快速计算与分流计算的目的

In [46]:
import multiprocessing
import time

haha_label = []
haha_label_one = []
haha_label_two = []
haha_label_three = []
haha_label_four = []
haha_label_five = []
haha_label_six = []
haha_label_seven = []
haha_label_eight = []
haha_label_night = []
haha_label_ten = []
haha_label_eleven = []
haha_label_twelve = []
haha_label_thirteen = []
haha_label_fourteen = []
haha_label_fifteen = []
haha_label_sixteen = []

def onehundred():
    for sent1 in haha['xdate'][0:100]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_one.append(math.fabs(preds[0]))
    one_list = pd.DataFrame(haha_label_one)
    one_list.to_excel('F:/label.xlsx', sheet_name='onehundred',index=False)

def twohundred():
    for sent1 in haha['xdate'][100:200]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_two.append(math.fabs(preds[0]))
    two_list = pd.DataFrame(haha_label_two)
    two_list.to_excel('F:/label.xlsx', sheet_name='twohundred',index=False)

def threehundred():
    for sent1 in haha['xdate'][200:300]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_three.append(math.fabs(preds[0]))
    three_list = pd.DataFrame(haha_label_three)
    three_list.to_excel('F:/label.xlsx', sheet_name='threehundred',index=False)

def fourhundred():
    for sent1 in haha['xdate'][300:400]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_four.append(math.fabs(preds[0])) 
    four_list = pd.DataFrame(haha_label_four)
    four_list.to_excel('F:/label.xlsx', sheet_name='fourhundred',index=False)

def fivehundred():
    for sent1 in haha['xdate'][400:500]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_five.append(math.fabs(preds[0])) 
    five_list = pd.DataFrame(haha_label_five)
    five_list.to_excel('F:/label.xlsx', sheet_name='fivehundred',index=False)

def sixhundred():
    for sent1 in haha['xdate'][500:600]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_six.append(math.fabs(preds[0])) 
    six_list = pd.DataFrame(haha_label_six)
    six_list.to_excel('F:/label.xlsx', sheet_name='sixhundred',index=False)

def sevenhundred():
    for sent1 in haha['xdate'][600:700]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_seven.append(math.fabs(preds[0])) 
    seven_list = pd.DataFrame(haha_label_seven)
    seven_list.to_excel('F:/label.xlsx', sheet_name='sevenhundred',index=False)

def eighthundred():
    for sent1 in haha['xdate'][700:800]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_eight.append(math.fabs(preds[0])) 
    eight_list = pd.DataFrame(haha_label_eight)
    eight_list.to_excel('F:/label.xlsx', sheet_name='eighthundred',index=False)

def ninehundred():
    for sent1 in haha['xdate'][800:900]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_night.append(math.fabs(preds[0])) 
    nine_list = pd.DataFrame(haha_label_night)
    nine_list.to_excel('F:/label.xlsx', sheet_name='ninehundred',index=False)

def tenhundred():
    for sent1 in haha['xdate'][900:1000]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_ten.append(math.fabs(preds[0])) 
    ten_list = pd.DataFrame(haha_label_ten)
    ten_list.to_excel('F:/label.xlsx', sheet_name='tenhundred',index=False)

def elevenhundred():
    for sent1 in haha['xdate'][1000:1100]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_eleven.append(math.fabs(preds[0])) 
    eleven_list = pd.DataFrame(haha_label_eleven)
    eleven_list.to_excel('F:/label.xlsx', sheet_name='elevenhundred',index=False)

def twelvehundred():
    for sent1 in haha['xdate'][1100:1200]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_twelve.append(math.fabs(preds[0])) 
    twelve_list = pd.DataFrame(haha_label_twelve)
    twelve_list.to_excel('F:/label.xlsx', sheet_name='twelvehundred',index=False)

def thirteenhundred():
    for sent1 in haha['xdate'][1200:1300]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_thirteen.append(math.fabs(preds[0])) 
    thirteen_list = pd.DataFrame(haha_label_thirteen)
    thirteen_list.to_excel('F:/label.xlsx', sheet_name='thirteenhundred',index=False)

def fourteenhundred():
    for sent1 in haha['xdate'][1300:1400]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_fourteen.append(math.fabs(preds[0])) 
    fourteen_list = pd.DataFrame(haha_label_fourteen)
    fourteen_list.to_excel('F:/label.xlsx', sheet_name='fourteenhundred',index=False)

def fifteenhundred():
    for sent1 in haha['xdate'][1400:1500]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_fifteen.append(math.fabs(preds[0])) 
    fifteen_list = pd.DataFrame(haha_label_fifteen)
    fifteen_list.to_excel('F:/label.xlsx', sheet_name='fifteenhundred',index=False)

def sixteenhundred():
    for sent1 in haha['xdate'][1500:1574]:
        for sent2 in haha['ydate']:
            sent1_ids = sent2index(vocab2id, jieba.lcut(sent1))
            sent2_ids = sent2index(vocab2id, jieba.lcut(sent2))
            sent1_pad = pad_sequences([sent1_ids], maxlen=max_len, padding='post')
            sent2_pad = pad_sequences([sent2_ids], maxlen=max_len, padding='post')
            preds = model.predict([sent1_pad, sent2_pad])
            haha_label_sixteen.append(math.fabs(preds[0])) 
    sixteen_list = pd.DataFrame(haha_label_sixteen)
    sixteen_list.to_excel('F:/label.xlsx', sheet_name='sixteenhundred',index=False)


if __name__ == '__main__':
    #target:指定函数名
    onehundred_process = multiprocessing.Process(target=onehundred)
    twohundred_process = multiprocessing.Process(target=twohundred)
    threehundred_process = multiprocessing.Process(target=threehundred)
    fourhundred_process = multiprocessing.Process(target=fourhundred)
    fivehundred_process = multiprocessing.Process(target=fivehundred)
    sixhundred_process = multiprocessing.Process(target=sixhundred)
    sevenhundred_process = multiprocessing.Process(target=sevenhundred)
    eighthundred_process = multiprocessing.Process(target=eighthundred)
    ninehundred_process = multiprocessing.Process(target=ninehundred)
    tenhundred_process = multiprocessing.Process(target=tenhundred)
    elevenhundred_process = multiprocessing.Process(target=elevenhundred)
    twelvehundred_process = multiprocessing.Process(target=twelvehundred)
    thirteenhundred_process = multiprocessing.Process(target=thirteenhundred)
    fourteenhundred_process = multiprocessing.Process(target=fourteenhundred)
    fifteenhundred_process = multiprocessing.Process(target=fifteenhundred)
    sixteenhundred_process = multiprocessing.Process(target=sixteenhundred)
 
    onehundred_process.start()
    twohundred_process.start()
    threehundred_process.start()
    fourhundred_process.start()
    fivehundred_process.start()
    sixhundred_process.start()
    sevenhundred_process.start()
    eighthundred_process.start()
    ninehundred_process.start()
    tenhundred_process.start()
    elevenhundred_process.start()
    twelvehundred_process.start()
    thirteenhundred_process.start()
    fourteenhundred_process.start()
    fifteenhundred_process.start()
    sixteenhundred_process.start()
# 因为jupyter notebook无法运行多进程会报错，以上代码在PyCharm中进行

### 通过十六个进程获取到相应的匹配度“label”的数据，并将数据保存至“label.xlsx”文件中

### 由于两百万条数据对于内核和内存运行的要求比较大，以下选取前600个求职者，即“600 * 1574 = 944,400” 大约一百万条数据去做模型匹配度分析

In [47]:
haha_label = pd.read_excel('F:/label.xlsx')
haha_label

label
0       0.562923
1       0.637812
2       0.552606
3       0.545203
4       0.616072
...          ...
944395  0.295034
944396  0.295034
944397  0.295034
944398  0.295034
944399  0.295034

[944400 rows x 1 columns]

In [48]:
len(haha_label)

944400

### 由于以下匹配度分析是每一个求职者对应每一个岗位数据，在列表中将前700个求职者的id与数据均遍历1574遍才能达到求职者信息与匹配度“label”的“1：1”关系，最后将数据保存

##### 依次遍历取得前600条数据的“工作id”，“岗位要求”，“人才id”，“人才求职需求”

In [50]:
haha_label_ax1 = []
for i in haha_x1[0:600]:
    for j in range(len(haha_x1)):
        haha_label_ax1.append(i)
len(haha_label_ax1) # 查看数据是否跟需要的944400条数据一致

944400

In [51]:
haha_label_ax2 = []
for i in look_work['工作id'][0:600]:
    for j in range(len(haha_x1)):
        haha_label_ax2.append(i)
len(haha_label_ax2)

944400

In [52]:
haha_label_ay1 = []
for z in haha_y1[0:600]:
    for g in range(len(haha_y1)):
        haha_label_ay1.append(z)
len(haha_label_ay1)

944400

In [53]:
haha_label_ay2 = []
for i in look_ability['人才id'][0:600]:
    for j in range(len(haha_x1)):
        haha_label_ay2.append(i)
len(haha_label_ay2)

944400

In [54]:
haha_label_axy1 = {
    '工作id':haha_label_ax2,
    '企业岗位要求':haha_label_ax1,
    '人才id':haha_label_ay2,
    '求职者岗位要求':haha_label_ay1,
    '岗位匹配度':haha_label
}
haha_label_data=pd.DataFrame(haha_label_axy1)
haha_label_data.to_csv('F:/resul3-1.csv',encoding='utf8')
haha_label_data

Unnamed: 0        工作id                   企业岗位要求        人才id  \
0                0  1.631110e+18       会计实习生3500.0经验不限  1.635610e+18   
1                1  1.631110e+18       会计实习生3500.0经验不限  1.635610e+18   
2                2  1.631110e+18       会计实习生3500.0经验不限  1.635610e+18   
3                3  1.631110e+18       会计实习生3500.0经验不限  1.635610e+18   
4                4  1.631110e+18       会计实习生3500.0经验不限  1.635610e+18   
...            ...           ...                            ...           ...   
944395      944395  1.482200e+18  大数据产品经理16000.0经验不限  7.537210e+18   
944396      944396  1.482200e+18  大数据产品经理16000.0经验不限  7.537210e+18   
944397      944397  1.482200e+18  大数据产品经理16000.0经验不限  7.537210e+18   
944398      944398  1.482200e+18  大数据产品经理16000.0经验不限  7.537210e+18   
944399      944399  1.482200e+18  大数据产品经理16000.0经验不限  7.537210e+18   

                                        求职者岗位要求     label  
0       数据分析师数据挖掘工程师算法工程师3000经验不限  0.562923  
1       数据分析师数据挖掘工程师算法工程师3000经验不限  0.637812  
2       数据分析师数据挖掘工程师算法工程师3000经验不限  0.552606  
3       数据分析师数据挖掘工程师算法工程师3000经验不限  0.545203  
4       数据分析师数据挖掘工程师算法工程师3000经验不限  0.616072  
...                                                ...       ...  
944395            数据分析师数据挖掘工程师8000经验不限  0.295034  
944396            数据分析师数据挖掘工程师8000经验不限  0.295034  
944397            数据分析师数据挖掘工程师8000经验不限  0.295034  
944398            数据分析师数据挖掘工程师8000经验不限  0.295034  
944399            数据分析师数据挖掘工程师8000经验不限  0.295034  

[944400 rows x 6 columns]

##### 为每条招聘信息提供岗位匹配度非0的求职者，将结果进行降序排序存放在“resul3-1.csv”文件中

In [55]:
haha_label_data.sort_values(by='label', inplace=True, ascending=False)
haha_label_data

Unnamed: 0        工作id  \
703606      703606  1.482200e+18   
782306      782306  1.482200e+18   
659534      659534  1.482200e+18   
736660      736660  1.482200e+18   
738234      738234  1.482200e+18   
...            ...           ...   
485096      485096  1.494560e+18   
485098      485098  1.494560e+18   
485085      485085  1.494560e+18   
485097      485097  1.494560e+18   
485083      485083  1.494560e+18   

                                                                     企业岗位要求  \
703606                     数据分析师/BI6000.01年工作经验                           
782306                          数据分析师10000.0经验不限                           
659534                       数据分析师10000.03年工作经验                           
736660     高级媒介专员（数据分析方向）10000.03年工作经验                           
738234                     数据产品经理15000.03年工作经验                           
...                                                   ...                           
485096  网络工程师/系统工程师（广州）6000.0经验不限系统测试平台管理数据通信......   
485098  网络工程师/系统工程师（广州）6000.0经验不限系统测试平台管理数据通信......   
485085  网络工程师/系统工程师（广州）6000.0经验不限系统测试平台管理数据通信......   
485097  网络工程师/系统工程师（广州）6000.0经验不限系统测试平台管理数据通信......   
485083  网络工程师/系统工程师（广州）6000.0经验不限系统测试平台管理数据通信......   

              人才id  \
703606  7.539530e+18   
782306  7.539480e+18   
659534  7.539530e+18   
736660  7.539480e+18   
738234  7.539480e+18   
...              ...   
485096  7.539740e+18   
485098  7.539740e+18   
485085  7.539740e+18   
485097  7.539740e+18   
485083  7.539740e+18   

                                                                                求职者岗位要求  \
703606  项目经理产品经理资质管理员电商大数据研究员5000经验不限......                             
782306  项目经理产品经理资质管理员电商大数据研究员4000经验不限......                             
659534  项目经理产品经理资质管理员电商大数据研究员5000经验不限......                             
736660               数据分析师数据挖掘工程师4000经验不限                                        
738234  项目经理产品经理资质管理员电商大数据研究员6000经验不限......                             
...                                                              ...                             
485096  数据分析师数据挖掘工程师6000经验不限数据探索数据预处理数据分析数据建模数据可视化......   
485098  数据分析师数据挖掘工程师6000经验不限数据探索数据预处理数据分析数据建模数据可视化......   
485085  数据分析师数据挖掘工程师6000经验不限数据探索数据预处理数据分析数据建模数据可视化......   
485097  数据分析师数据挖掘工程师6000经验不限数据探索数据预处理数据分析数据建模数据可视化......   
485083  数据分析师数据挖掘工程师6000经验不限数据探索数据预处理数据分析数据建模数据可视化......   

           label  
703606  0.867432  
782306  0.867432  
659534  0.862260  
736660  0.862260  
738234  0.862260  
...          ...  
485096  0.005637  
485098  0.005008  
485085  0.005008  
485097  0.005008  
485083  0.005008  

[944400 rows x 6 columns]

In [56]:
haha_label_data.to_csv('F:/amyself/c题爬虫/第三题/resul3-1.csv')